File: py_anc/algorithms/mgdfxlms.py
from __future__ import annotations

from typing import Dict

import numpy as np
from scipy.linalg import convolution_matrix


def mgdfxlms(params: Dict):
    """Distributed MGDFxLMS algorithm."""
    time = params["time"]
    rir_manager = params["rir_manager"]
    network = params["network"]
    l = int(params["L"])
    x = np.asarray(params["reference_signal"], dtype=float)
    d = np.asarray(params["desired_signal"], dtype=float)

    key_ref_mics = list(rir_manager.reference_microphones.keys())
    key_sec_spks = list(rir_manager.secondary_speakers.keys())
    key_err_mics = list(rir_manager.error_microphones.keys())

    ref_idx = {rid: i for i, rid in enumerate(key_ref_mics)}
    sec_idx = {sid: i for i, sid in enumerate(key_sec_spks)}
    err_idx = {eid: i for i, eid in enumerate(key_err_mics)}

    num_ref_mics = len(key_ref_mics)
    num_sec_spks = len(key_sec_spks)
    num_err_mics = len(key_err_mics)

    if x.shape[1] != num_ref_mics:
        raise ValueError(
            f"reference_signal columns ({x.shape[1]}) must match number of reference microphones ({num_ref_mics})."
        )

    n_samples = len(time)

    max_ls_hat = 0
    for sec_id in key_sec_spks:
        ls_hat = len(rir_manager.get_secondary_rir(sec_id, key_err_mics[0]))
        if ls_hat > max_ls_hat:
            max_ls_hat = ls_hat

    x_taps = np.zeros((max(l, max_ls_hat), num_ref_mics), dtype=float)
    e = np.zeros((n_samples, num_err_mics), dtype=float)

    y_taps = []
    for sec_id in key_sec_spks:
        sec_len = len(rir_manager.get_secondary_rir(sec_id, key_err_mics[0]))
        y_taps.append(np.zeros(sec_len, dtype=float))

    for node in network.nodes.values():
        if node.ref_mic_id not in ref_idx:
            raise ValueError(f"Node {node.node_id} has invalid ref_mic_id={node.ref_mic_id}")
        node.init(l)

    print("预计算交叉路径补偿滤波器 (ckm)...")
    for node in network.nodes.values():
        for neighbor_id in node.neighbor_ids:
            if neighbor_id == node.node_id:
                continue
            neighbor = network.nodes[neighbor_id]

            s_hat_km = np.asarray(rir_manager.get_secondary_rir(node.sec_spk_id, neighbor.err_mic_id), dtype=float)
            s_hat_mm = np.asarray(rir_manager.get_secondary_rir(neighbor.sec_spk_id, neighbor.err_mic_id), dtype=float)

            mtx = convolution_matrix(s_hat_mm, node.lc, mode="full")
            expected_len = mtx.shape[0]
            current_len = len(s_hat_km)

            if current_len < expected_len:
                b = np.concatenate([s_hat_km, np.zeros(expected_len - current_len, dtype=float)])
            elif current_len > expected_len:
                b = s_hat_km[:expected_len]
                print(
                    f"Warning: Truncating S_hat_km from length {current_len} to {expected_len} "
                    f"for node {node.node_id} -> neighbor {neighbor_id}"
                )
            else:
                b = s_hat_km

            ckm, *_ = np.linalg.lstsq(mtx, b, rcond=None)
            node.ckm_taps[neighbor_id] = ckm

            s_km_approx = np.convolve(s_hat_mm, ckm)[: len(s_hat_km)]
            rel_err = np.linalg.norm(s_km_approx - s_hat_km) / (np.linalg.norm(s_hat_km) + 1e-12)
            print(
                f"Node {node.node_id} -> Neighbor {neighbor_id}: 相对误差 = {rel_err:.4f} ({rel_err * 100:.1f}%)"
            )

    print("开始MGDFxLMS仿真...")
    for n in range(n_samples):
        x_taps = np.vstack([x[n : n + 1, :], x_taps[:-1, :]])

        for node in network.nodes.values():
            ref_col = ref_idx[node.ref_mic_id]
            y = float(node.w @ x_taps[:l, ref_col])
            k = sec_idx[node.sec_spk_id]
            y_taps[k] = np.concatenate(([y], y_taps[k][:-1]))

        for m, err_id in enumerate(key_err_mics):
            yf = 0.0
            for k, sec_id in enumerate(key_sec_spks):
                s = rir_manager.get_secondary_rir(sec_id, err_id)
                ls = len(s)
                yf += float(np.dot(s, y_taps[k][:ls]))
            e[n, m] = d[n, m] + yf

        for node in network.nodes.values():
            ref_col = ref_idx[node.ref_mic_id]
            s_hat = rir_manager.get_secondary_rir(node.sec_spk_id, node.err_mic_id)
            ls_hat = len(s_hat)
            xf = float(s_hat @ x_taps[:ls_hat, ref_col])
            node.xf_taps = np.concatenate(([xf], node.xf_taps[:-1]))
            node.gradient = e[n, err_idx[node.err_mic_id]] * node.xf_taps
            node.direction = node.gradient.copy()

        for node in network.nodes.values():
            for neighbor_id in node.neighbor_ids:
                if neighbor_id == node.node_id:
                    continue
                neighbor = network.nodes[neighbor_id]

                ckm = node.ckm_taps[neighbor.node_id]
                ckm_flipped = ckm[::-1]
                full_conv = np.convolve(neighbor.gradient, ckm_flipped)
                start = node.lc - 1
                correction = full_conv[start : start + l]

                if correction.shape[0] < l:
                    correction = np.pad(correction, (0, l - correction.shape[0]))

                node.direction = node.direction + correction

        for node in network.nodes.values():
            node.w = node.w - node.step_size * node.direction

    filter_coeffs = {node.sec_spk_id: node.w.copy() for node in network.nodes.values()}
    return {"err_hist": e, "filter_coeffs": filter_coeffs}

File: matlab_impl/+algorithms/MGDFxLMS.m
function results = MGDFxLMS(params)
% MGDFxLMS 分布式主动噪声控制FxLMS算法
% 描述:
%
% 输入:
%   params: 包含所有仿真参数的结构体。
%       - params.time: 离散时间轴
%       - params.rirManager: RIRManager对象，用于获取脉冲响应
%       - params.network: 通讯网络拓扑结构
%       - params.L: 控制滤波器的长度
%       - params.referenceSignal: 参考信号
%       - params.desiredSignal: 期望信号
%
% 输出:
%   results: 包含仿真结果的结构体。
%       - results.err_hist: 麦克风处的误差信号历史
%       - results.filter_coeffs: 各节点最终的控制器滤波器系数
%% 1. 解包参数
time            = params.time;
rirManager      = params.rirManager;
network         = params.network;
L               = params.L;
x               = params.referenceSignal;
d               = params.desiredSignal;

% 从rirManager获取参数
keyRefMics = keys(rirManager.ReferenceMicrophones);
keySecSpks = keys(rirManager.SecondarySpeakers);
keyErrMics = keys(rirManager.ErrorMicrophones);

numRefMics      = numEntries(rirManager.ReferenceMicrophones);
numSecSpks      = numEntries(rirManager.SecondarySpeakers);
numErrMics      = numEntries(rirManager.ErrorMicrophones);

if size(x, 2) ~= numRefMics
    error('referenceSignal columns (%d) must match number of reference microphones (%d).', size(x, 2), numRefMics);
end

nSamples = length(time);

%% 2. 初始化
max_Ls_hat = 0;
for i = keySecSpks'
    Ls_hat = length(rirManager.getSecondaryRIR(i, keyErrMics(1)));
    if Ls_hat > max_Ls_hat
        max_Ls_hat = Ls_hat;
    end
end

x_taps = zeros(max([L, max_Ls_hat]), numRefMics);

e = zeros(nSamples, numErrMics); % 误差信号

y_taps = cell(numSecSpks);   % 控制信号
for k = 1:numSecSpks
    y_taps{k} = zeros(length(rirManager.getSecondaryRIR(keySecSpks(k), keyErrMics(1))), 1);
end

for keyNode = keys(network.Nodes)'
    node = network.Nodes(keyNode);
    node.init(L);
end

fprintf('预计算交叉路径补偿滤波器 (ckm)...\n');
for keyNode = keys(network.Nodes)'
    node = network.Nodes(keyNode);
    for neighborId = node.NeighborIds
        if neighborId == node.Id
            continue
        end
        neighbor = network.Nodes(neighborId);

        % 2.1 计算ckm 可能能优化
        % 获取 RIR 并强制转换为列向量 (:)
        S_hat_km = rirManager.getSecondaryRIR(node.SecSpkId, neighbor.ErrMicId);
        S_hat_km = S_hat_km(:);

        S_hat_mm = rirManager.getSecondaryRIR(neighbor.SecSpkId, neighbor.ErrMicId);
        S_hat_mm = S_hat_mm(:);

        % 生成卷积矩阵
        M = convmtx(S_hat_mm, node.Lc);

        expected_len = size(M, 1);
        current_len = length(S_hat_km);

        % 构造目标向量 b (列向量)
        if current_len < expected_len
            % 补零
            b = [S_hat_km; zeros(expected_len - current_len, 1)];
        elseif current_len > expected_len
            % 截断 不应该发生
            b = S_hat_km(1:expected_len);
            fprintf("Warning: Truncating S_hat_km from length %d to %d for node %d -> neighbor %d\n", ...
                current_len, expected_len, node.Id, neighborId);
        else
            b = S_hat_km;
        end
        % 最小二乘求解
        ckm = M \ b;
        % fprintf('Node %d -> Neighbor %d: ckm norm = %.4f, max = %.4f\n', ...
        %        node.Id, neighborId, norm(ckm), max(abs(ckm)));
        node.ckm_taps(neighborId) = {ckm};

        % 2.2 验证ckm近似质量
        % 计算近似值
        S_km_approx = conv(S_hat_mm, ckm);
        S_km_approx = S_km_approx(1:length(S_hat_km));

        % 计算误差
        err = norm(S_km_approx - S_hat_km) / norm(S_hat_km);
        fprintf('Node %d -> Neighbor %d: 相对误差 = %.4f (%.1f%%)\n', ...
            node.Id, neighborId, err, err*100);
    end
end

%% 3. 主循环
fprintf('开始MGDFxLMS仿真... \n');
for n = 1:nSamples
    % 3.1. 更新参考信号状态向量 (全局)
    x_taps = [x(n, :); x_taps(1:end-1, :)];

    % 3.2. 生成控制信号 y(n) (分布式)
    for keyNode = keys(network.Nodes)'
        node = network.Nodes(keyNode);
        y = node.W' * x_taps(1:L, keyRefMics == node.RefMicId);
        y_taps{keySecSpks == node.SecSpkId} = [y; y_taps{keySecSpks == node.SecSpkId}(1:end-1)];
    end

    % 3.3. 计算误差信号 e(n) (全局)
    for m = 1:numErrMics
        yf = 0;
        for k = 1:numSecSpks
            S = rirManager.getSecondaryRIR(keySecSpks(k), keyErrMics(m));
            Ls = length(S);
            yf = yf + S * y_taps{k}(1:Ls);
        end
        e(n, m) = d(n, m) + yf;
    end

    % 3.4. 更新节点信息 (分布式)
    % 3.4.1 计算xf、梯度
    for keyNode = keys(network.Nodes)'
        node = network.Nodes(keyNode);
        S_hat = rirManager.getSecondaryRIR(node.SecSpkId, node.ErrMicId);
        Ls_hat = length(S_hat);
        xf = S_hat * x_taps(1:Ls_hat, keyRefMics == node.RefMicId);
        node.xf_taps = [xf; node.xf_taps(1:end-1)];
        node.gradient = e(n, keyErrMics == node.ErrMicId) * node.xf_taps;
        node.direction = e(n, keyErrMics == node.ErrMicId) * node.xf_taps;
    end
    % 3.4.2 收集所有邻居的梯度信息，并乘以ckm进行补偿
    for keyNode = keys(network.Nodes)'
        node = network.Nodes(keyNode);
        for neighborId = node.NeighborIds
            if neighborId == node.Id
                continue;
            end
            neighbor = network.Nodes(neighborId);

            ckm_cell = node.ckm_taps(neighbor.Id);
            ckm_vec = ckm_cell{1};

            ckm_flipped = flipud(ckm_vec);
            full_conv = conv(neighbor.gradient, ckm_flipped);
            % 截断（取“有效”部分）
            % 注意：由于翻转了，有效数据的起始点变了
            valid_start = node.Lc;
            % 提取正确的部分 (长度为 N)
            correction = full_conv(valid_start : valid_start + L - 1);

            node.direction = node.direction + correction;
        end
    end
    % 3.4.3 更新W
    for keyNode = keys(network.Nodes)'
        node = network.Nodes(keyNode);
        node.W = node.W - node.StepSize * node.direction;
    end
end

%% 4. 打包结果
filter_coeffs = dictionary;
for keyNode = keys(network.Nodes)'
    node = network.Nodes(keyNode);
    w = node.W;
    filter_coeffs(node.SecSpkId) = {w};
end
results.err_hist      = e;
results.filter_coeffs = filter_coeffs;
end